Querybuilder: add json children support to where clauses

ryanmitchell opened the pull request:
Ok last one I promise... no tests with this one as I just wanted to see if this is how you would approach it. It does work, but I'm convinced there must be a better way.

This lets you query "inside" fields (JSON where - https://laravel.com/docs/8.x/queries#json-where-clauses)

```
Entry::query()
->where('replicator_field->0->sub_field', 'value')
->get()
```

ryanmitchell commented:
I've added some tests for EntryQueryBuilder to prove this concept works. I'd love to hear your thoughts on this approach and if there is a better way that would also allow it to be supported by IteratorQueryBuilder.

jasonvarga commented:
I don't know the answer to your approach question, but I wanted to just drop this in here.

It seems this this PR would be super related to https://github.com/statamic/ideas/issues/428. Not suggesting you make any changes, but really just a note for when we do a thorough review.

ryanmitchell commented:
Yeah sorting would be something that should be covered by this (it should already be - haven't tested to be sure). My approach question was really "is this how you would do this Jason"?

jasonvarga commented:
It makes sense to me. I would think to get it to work on the iterator builder you'd need to duplicate the getItemPartValue stuff.

ryanmitchell commented:
I've added support for IteratorBuilder alongside AssetBuilder tests and UserBuilder tests. It all seems to work well at this point.

With reference to sorting in [statamic/ideas#428](https://github.com/statamic/ideas/issues/428), I think you will need to use a custom sortBy closure... eg. 

```
            ->sortByDesc(function ($entry) {
                return array_get($entry->data(), Carbon::parse('date_range_field.start'));
            })
```

Unless you know of a better way?

jasonvarga commented:
As for sorting, this PR makes it work without any special additions for sorting. Nice!

In a template you can do this:

```
{{ collection:blog sort="foo->bar:desc" }}
```

However, we don't use the json arrow syntax anywhere else in Antlers, so that's a little weird. In a separate PR I'll make it work with the colon syntax, which will probably be what people instinctively reach for:

```
sort="foo:bar:desc"
```

ryanmitchell commented on src/Stache/Indexes/Value.php:
```diff
@@ -30,14 +45,26 @@ public function getItemValue($item)
+            return $item[$name] ?? null;
+        }
+
+        if (is_scalar($item)) {
+            return null;
+        }
+
+        if ($item instanceof stdClass) {
+            return $item->$name ?? null;
+        }
```
`is_scalar` covers the situation where you attempt to get a child (bar) from a parent (foo) that doesn't have children. Otherwise it was throwing an error for me when testing.

`stdClass` I'll remove - I was just trying to be thorough and cover anything I thought people might throw at it.

ryanmitchell commented:
Amazing - glad that sorting works so simply.

jasonvarga commented:
I extracted the item part stuff into its own class so it lives in one spot.

I also added tests for the scalar handling. I checked in MySQL what would happen if you queried a JSON field that was a string, and it basically just ignores it like you've done. 👍 

Thanks for this!

ryanmitchell commented:
yeah I saw that - I should have thought to do it myself.

No problem! 



> On 7 Jan 2022, at 17:59, Jason Varga ***@***.***> wrote:
> 
> ﻿
> I extracted the item part stuff into its own class so it lives in one spot.
> 
> I also added tests for the scalar handling. I checked in MySQL what would happen if you queried a JSON field that was a string, and it basically just ignores it like you've done. 👍
> 
> Thanks for this!
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you authored the thread.
